presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -467,18 +488,28 @@ public void processScheduleResults(StageState newState, Set<RemoteTask> newTasks
 
             if (parent.isPresent()) {
                 // Add an exchange location to the parent stage for each new task
-                Set<URI> newExchangeLocations = newTasks.stream()
-                        .map(task -> task.getTaskInfo().getSelf())
-                        .collect(toImmutableSet());
-                parent.get().addExchangeLocations(currentStageFragmentId, newExchangeLocations, noMoreTasks);
+                ImmutableList.Builder<BiFunction<TaskId, Integer, URI>> newExchangeLocations = ImmutableList.builder();
+                for (RemoteTask task : newTasks) {
+                    // if this stage has a broadcast output buffer...
+                    URI taskLocation = task.getTaskInfo().getSelf();
+                    if (currentStageIsBroadcast) {
```

Comment:
A possible alternative to this:

* Keep both task id and partition id in the URL
* Let the OutputBuffer pick the one (of task id and partition id) it is interested in and throw out the other one